Change the discovery to use a fingerprint DNS resolvable name rather than an IP address

samgiles opened the pull request:
This needs to land around the same time as https://github.com/fxbox/registration_server/pull/18

samgiles commented:
r? @azasypkin This works with the changes currently in the registration server at `master` and https://github.com/fxbox/foxbox/pull/281

azasypkin commented on app/js/lib/foxbox/foxbox.js:
```diff
@@ -97,16 +97,16 @@ export default class Foxbox extends Service {
   /**
-   * Get the IP address of the box on the local network using the registration
+   * Get the URL of the box using the registration
    * server.
```
__nit:__ please move ```server.``` to the line above as there is enough space for it now :)

azasypkin commented on app/js/lib/foxbox/foxbox.js:
```diff
@@ -134,6 +134,16 @@ export default class Foxbox extends Service {
             boxes
               .filter(box => box.timestamp - now >= 0)
+              .map(box => {
+
```
__nit:__ redundant empty line

azasypkin commented on app/js/lib/foxbox/foxbox.js:
```diff
@@ -134,6 +134,16 @@ export default class Foxbox extends Service {
+                  // NOTE(sgiles): There is consideration to allow
+                  // only "local_url" and "tunnel_url", removing the need
+                  // to parse message
+                  let message = JSON.parse(box.message);
```
__nit:__ ```const {local_url, tunnel_url} = JSON.parse(box.message);```

azasypkin commented on app/js/lib/foxbox/foxbox.js:
```diff
@@ -134,6 +134,16 @@ export default class Foxbox extends Service {
+                  let message = JSON.parse(box.message);
+                  box.local_url  = message.local_url;
+                  box.tunnel_url = message.tunnel_url;
+                  return box;
```
__optional nit:__ I think it's OK to freeze ```box``` here to avoid one more ```map``` iteration.

azasypkin commented on app/js/lib/foxbox/network.js:
```diff
@@ -56,29 +56,27 @@ export default class Network {
   }
 
-  get localOrigin() {
+  get localUrl() {
```
__question:__ Why do we rename ```origin``` to ```URL```? I thought ```local_url``` and ```tunnel_url``` are always just ```origins``` [1] or they can contain any ```path``` part as well? It just looks a bit confusing when we create URLs with something like ```${box.local_url}\path1\path2``` - ```box.local_url``` looks more like ```box.local_origin``` or at least ```box.local_baseURL```

[1] https://tools.ietf.org/html/rfc6454#section-3.2

azasypkin commented on app/js/lib/foxbox/settings.js:
```diff
@@ -83,64 +76,24 @@ export default class Settings extends Model {
-    storage.setItem(`${PREFIX}tunnelHostname`, this._tunnelHostname);
+  set localUrl(url) {
+    url = String(url) || DEFAULT_URL;
+    this._localUrl = url.replace(/\/$/, ''); // Trailing slash.
```
If we agree that it's ```origin``` and not ```url``` we can do just ```this._localOrigin = url ? (new URL(url)).origin : DEFAULT_ORIGIN;``` that will guard us from non-valid URL's (at least at development stage) and get rid of trailing slashes :)

azasypkin commented on app/js/views/user-login.jsx:
```diff
@@ -31,13 +31,13 @@ export default class UserLogin extends React.Component {
       let value = this.state.value !== null ? 0 : this.state.value;
       const options = this.state.boxes.map((box, index) => {
-        if (box.local_ip === this.foxbox.localHostname) {
+        if (box.client === this.foxbox.client) {
```
__possible issue:__ what is the ```client``` field? I don't see it anywhere else (e.g. in ```this.foxbox```)?

samgiles replied:
This is a field that gets returned from the registration server.  It's the fingerprint of the self-signed certificate the box generates.

azasypkin replied:
Yeah, but could you point me to the place where ```this.foxbox.client``` is initialized?

samgiles replied:
:D, I see what you meant now.  I've added `clientId` to `foxbox`.

azasypkin commented on app/js/views/user-login.jsx:
```diff
@@ -31,13 +31,13 @@ export default class UserLogin extends React.Component {
-        const label = box.local_ip +
-          (box.tunnel_url ? ` / ${box.tunnel_url}` : '');
+
+        const label = box.client;
```
__nit:__ I'd get rid of ```label``` variable and use ```box.client``` directly in JSX

azasypkin commented:
Overall changes look good to me, I've just left few nits, one question (I'm not sure if we should rename ```origin``` to ```URL```) and mentioned one possible issue (unknown and suspicious ```this.foxbox.client``` field).

Thanks!

samgiles commented:
I agree with the `origin`/`url` change, and I've changed the name in https://github.com/fxbox/foxbox/pull/281 to reflect that it should be `origin`. :) thanks, I'll update.

samgiles commented:
Hey @azasypkin, thanks for the review, I've addressed your nits. Reverted back to the `origin` name, and added clientId to the `foxbox` object.  How does it look now?

samgiles commented:
Rebased and squashed

gmarty commented on app/js/lib/foxbox/network.js:
```diff
@@ -56,7 +56,7 @@ export default class Network {
   }
 
-  get origin() {
+  get url() {
```
Why renaming `origin` by `url`? AFAIK a URL is something like `example.com`, while an origin is `https://example.com:443`.

samgiles replied:
:grimacing: I missed these - it should be `origin`.

samgiles commented:
Just realised I may have merged something that was removed from master back in.

gmarty commented on app/js/lib/foxbox/settings.js:
```diff
@@ -5,14 +5,12 @@ import { Model } from 'components/mvc';
 const ONLINE_CHECKING_INTERVAL = 5000;
 const ONLINE_CHECKING_LONG_INTERVAL = 1000 * 60 * 5;
-const REGISTRATION_SERVICE = 'http://knilxof.org:4242/ping';
+const REGISTRATION_SERVICE = 'http://localhost:4242/ping';
```
Are we not using the registration server anymore?

gmarty commented on app/js/lib/foxbox/foxbox.js:
```diff
@@ -134,7 +137,17 @@ export default class Foxbox extends Service {
               .filter(box => box.timestamp - now >= 0)
-              .map(box => Object.freeze(box))
+              .map(box => {
+
```
nit: line break not needed.

gmarty commented on app/js/lib/foxbox/foxbox.js:
```diff
@@ -134,7 +137,17 @@ export default class Foxbox extends Service {
+                  // only "local_origin" and "tunnel_origin", removing the need
+                  // to parse message - this merges the relevant message fields
+                  // into the main object
+                  const {local_origin, tunnel_origin} = JSON.parse(box.message);
```
nit: we usually add 1 space within braces: `{ local_origin, tunnel_origin }`.

gmarty commented:
I left a couple of nits but the rest looks good to me.
r+!
